### PR TITLE
doc: build: dts: Remove refs to label property

### PR DESCRIPTION
**Pull request by microbuilder**
Removes references to the `label` property in devicetree, replacing
them with node labels where appropriate.

Signed-off-by: Kevin Townsend <kevin.townsend@linaro.org>

### CODE REVIEW

File: doc/build/dts/bindings.rst [hunk 1]
```diff
@@ -298,20 +298,15 @@ For example, a binding for a UART peripheral might look something like this:
...
-       type: string
-       description: human-readable name
-       required: false
 
 The properties in the following node would be validated by the above binding:
 
 .. code-block:: devicetree
 
-   my-serial@deadbeef {
+   uart_0:my-serial@deadbeef {
```
Comment (mbolivar-nordic) [inline doc/build/dts/bindings.rst:306]:
```suggestion
   uart_0: my-serial@deadbeef {
```

File: doc/build/dts/howtos.rst [hunk 2]
```diff
@@ -122,21 +122,18 @@ that the device is ready to be used before passing it to any API functions.
...
+      device_get_binding(DEVICE_DT_NAME(MY_SERIAL));
 
 You can then use ``uart_dev`` with :ref:`uart_api` API functions like
 :c:func:`uart_configure`. Similar code will work for other device types; just
 make sure you use the correct API for the device.
 
-There's no need to override the ``label`` property to something else: just make
-a node identifier and pass it to ``DT_LABEL`` to get the right string to pass
-to ``device_get_binding()``.
-
```
Comment (mbolivar-nordic) [inline doc/build/dts/howtos.rst:139]: Looking at it again, this whole example doesn't really make sense to me. If the device name comes from user input, then we aren't getting it from a DT node at all. Just deleting this part completely makes more sense to me right now.

File: doc/build/dts/intro.rst [hunk 3]
```diff
@@ -166,24 +162,20 @@ This is the corresponding DTS:
 
    / {
            soc {
-                   i2c@40003000 {
+                   i2c_0:i2c@40003000 {
```
Comment (mbolivar-nordic) [inline doc/build/dts/intro.rst:165]: Do you really need the node labels here? It looks like they aren't used. They complicate the example, so since they are not required and users have already been told that zero is a legal number of node labels, I don't think adding node labels here makes the documentation clearer.

**Review by mbolivar-nordic — COMMENTED** [review]
Looks good, just a few minor things.

I think we should probably mention this effort at the TSC along with this pull request, to try to maximize awareness, before merging.

There are also still a ton of label properties throughout the tree, so I'm not sure if we want to merge this now, or wait until the majority are gone before proceeding.

What was your thinking @microbuilder ?

File: doc/build/dts/howtos.rst (hunk 2 above)
Comment (microbuilder) [reply doc/build/dts/howtos.rst]: Removed, but I'll leave this one open in case there are other opinions.

**Comment by microbuilder** [comment]
> Looks good, just a few minor things.
> 
> I think we should probably mention this effort at the TSC along with this pull request, to try to maximize awareness, before merging.
> 
> There are also still a ton of label properties throughout the tree, so I'm not sure if we want to merge this now, or wait until the majority are gone before proceeding.
> 
> What was your thinking @microbuilder ?

Something we mentioned on the dev review call was adding a section describing historical usage of the label property, and why it shouldn't be used moving forward, but the reason it exists today in the code base.

That might be enough to merge this in and make some progress to encourage node labels not label properties, without waiting on all the drivers to be updated.

I'll put a draft together and push it here.

File: doc/build/dts/bindings.rst [hunk 4]
```diff
@@ -298,20 +298,15 @@ For example, a binding for a UART peripheral might look something like this:
...
-       type: string
-       description: human-readable name
-       required: false
 
 The properties in the following node would be validated by the above binding:
 
 .. code-block:: devicetree
 
-   my-serial@deadbeef {
+   uart_0: my-serial@deadbeef {
```
Comment (galak) [inline doc/build/dts/bindings.rst:306]: maybe drop the nodelabel for now.  Not seeing it used/referenced.

File: doc/build/dts/howtos.rst [hunk 5]
```diff
@@ -120,23 +120,6 @@ device may not be initialized, or may have failed to initialize, you must verify
 that the device is ready to be used before passing it to any API functions.
 (This check is done for you by :c:func:`device_get_binding`.)
 
-In some situations the device cannot be known at build-time, e.g., if it depends
```
Comment (galak) [inline doc/build/dts/howtos.rst:123]: keep this section, its still relevant.  Change `device_get_binding(DT_LABEL(MY_SERIAL))` to device_get_binding(DEVICE_DT_NAME(MY_SERIAL));`

**Review by galak — CHANGES_REQUESTED** [review]
Some changes, and probably good to add a historical section on `label` somewhere.

File: doc/build/dts/intro.rst [hunk 6]
```diff
@@ -350,6 +328,15 @@ interrupts
...
 
+.. note::
+   
+   Earlier versions of Zephyr made frequent use of the ``label`` property,
+   which is distinct from the standard :ref:`node label <dt-node-labels>`.
+   Use of the label *property* and the corresponding :c:func:`DT_LABEL` macro is
+   actively discouraged in new devicetree bindings. Label properties continue to
+   persist for historical reasons in some existing bindings and overlays, but
+   should not be used in new bindings or device implementations.
+
```
Comment (microbuilder) [inline doc/build/dts/intro.rst:339]: @mbolivar-nordic @galak Sensible sounding?

File: doc/build/dts/howtos.rst (hunk 5 above)
Comment (microbuilder) [reply doc/build/dts/howtos.rst:123]: Added back.

**Review by henrikbrixandersen — DISMISSED** [review]

File: doc/build/dts/intro.rst (hunk 6 above)
Comment (mbolivar-nordic) [reply doc/build/dts/intro.rst]: DT_LABEL isn't used in bindings, and the right role is `:c:macro:`, so I would reword to:

```
Use of the label *property* in new devicetree bindings, as well as use of the :c:macro:`DT_LABEL` in new code, is actively discouraged.
```

Comment (microbuilder) [reply doc/build/dts/intro.rst]: Fixed, thanks.

**Review by henrikbrixandersen — DISMISSED** [review]

File: doc/build/dts/howtos.rst [hunk 7]
```diff
@@ -122,21 +122,18 @@ that the device is ready to be used before passing it to any API functions.
 
 In some situations the device cannot be known at build-time, e.g., if it depends
 on user input like in a shell application. In this case you can get the
-``struct device`` by combining :c:func:`DT_LABEL` with
+``struct device`` by combining :c:func:`DEVICE_DT_NAME` with
 :c:func:`device_get_binding`:
```
Comment (gmarull) [inline doc/build/dts/howtos.rst:126]: `DEVICE_DT_NAME` doesn't really help if we depend on user input, so this paragraph isn't really helpful as it is written now.

**Comment by microbuilder** [comment]
There was an svg file I missed with the label property ... updates the two image files to remove this.

**Comment by microbuilder** [comment]
@mbolivar-nordic @galak @gmarull Any consensus on what to put in the case of devices coming from user input?

**Comment by gmarull** [comment]
> @mbolivar-nordic @galak @gmarull Any consensus on what to put in the case of devices coming from user input?

That's a good question... The main problem I see with label removal is that shells like https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pwm/pwm_shell.c will suddenly become unusable because instead of providing `PWM_0` as an input (easy to remember, write) I now need to provide `pwm@4001c000`. If we had device classes, shells could at least tell you "hey, this is the list of available PWM devices I can work with, pick one".

**Comment by microbuilder** [comment]
> > @mbolivar-nordic @galak @gmarull Any consensus on what to put in the case of devices coming from user input?
> 
> That's a good question... The main problem I see with label removal is that shells like https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pwm/pwm_shell.c will suddenly become unusable because instead of providing `PWM_0` as an input (easy to remember, write) I now need to provide `pwm@4001c000`. If we had device classes, shells could at least tell you "hey, this is the list of available PWM devices I can work with, pick one".

Good point, similar usability issue with sensor_shell.

**Comment by mbolivar-nordic** [comment]
@gmarull that's a fair point, but it seems like a separate issue from what this PR is trying to solve.

I'm glad I'm not alone in thinking that this content doesn't make sense as is, though.

Why not something like this?

```suggestion
   const char *dev_name = /* read device name from user */;
   const struct device *uart_dev = device_get_binding(dev_name);
```

**Comment by microbuilder** [comment]
> Some changes, and probably good to add a historical section on `label` somewhere.

@galak Mind having another look?

**Comment by microbuilder** [comment]
Coding guidelines failing due to [launchpadlibrarian.com](https://launchpadlibrarian.net/) being down:

```
 Processing triggers for man-db (2.9.1-1) ...
--2022-08-11 13:30:34--  https://launchpad.net/~npalix/+archive/ubuntu/coccinelle/+files/coccinelle_1.0.8~20.04npalix1_amd64.deb
Resolving launchpad.net (launchpad.net)... 185.125.189.222, 185.125.189.223, 2620:2d:4000:1001::8004, ...
Connecting to launchpad.net (launchpad.net)|185.125.189.222|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://launchpadlibrarian.net/484859880/coccinelle_1.0.8~20.04npalix1_amd64.deb [following]
--2022-08-11 13:30:35--  https://launchpadlibrarian.net/484859880/coccinelle_1.0.8~20.04npalix1_amd64.deb
Resolving launchpadlibrarian.net (launchpadlibrarian.net)... 185.125.189.229, 185.125.189.228, 2620:2d:4000:1001::8007, ...
Connecting to launchpadlibrarian.net (launchpadlibrarian.net)|185.125.189.229|:443... connected.
HTTP request sent, awaiting response... 503 Service Unavailable
2022-08-11 13:30:36 ERROR 503: Service Unavailable.
```

File: doc/build/dts/howtos.rst (hunk 7 above)
Comment (microbuilder) [reply doc/build/dts/howtos.rst]: Updates

**Review by gmarull — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

File: doc/build/dts/howtos.rst [hunk 8]
```diff
@@ -122,21 +122,18 @@ that the device is ready to be used before passing it to any API functions.
...
 
 .. code-block:: c
 
-   const struct device *uart_dev = device_get_binding(DT_LABEL(MY_SERIAL));
+   const char *dev_name = /* TODO: insert device name from user */;
+   const struct device *uart_dev = device_get_binding(dev_name);
```
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply doc/build/dts/howtos.rst:131]: > Acceptable as-is?

:+1:

**Review by mbolivar-nordic — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
lgtm

**Review by galak — APPROVED** [review]
